prairielib/lib/migrations.js

```diff
@@ -34,136 +35,272 @@ module.exports.init = function (dir, proj, callback) {
-        sqldb.queryOneRow(sql.get_last_migration, { project }, (err, results) => {
-          if (ERR(err, callback)) return;
-          let last_migration = results.rows[0].last_migration;
-          if (last_migration == null) {
-            last_migration = -1;
-            noExistingMigrations = true;
+        }
+        try {
+          await sqldb.queryAsync('SELECT timestamp FROM migrations;', {});
+        } catch (err) {
+          if (err.routine === 'errorMissingColumn') {
+            logger.info('Altering migrations table again');
+            await sqldb.queryAsync(sql.add_timestamp_column, {});
+          } else {
+            throw err;
           }
-          callback(null, last_migration);
-        });
-      },
-      (last_migration, callback) => {
-        fs.readdir(migrationDir, (err, files) => {
-          if (ERR(err, callback)) return;
-
-          const regex = /^([0-9]+)_.+\.sql$/;
-          files = files
-            .filter((file) => regex.test(file))
-            .map((file) => {
-              const index = Number.parseInt(regex.exec(file)[1]);
-              return {
-                index: index,
-                filename: file,
-              };
-            })
-            .sort((a, b) => {
-              return a.index - b.index;
-            });
-
-          // check that we don't have repeated indexes
-          const repeatedIndexes = _(files)
-            .groupBy((file) => file.index)
-            .pickBy((fileList) => fileList.length > 1)
-            .map(
-              (fileList, index) =>
-                `Repeated index ${index}:\n` + _.map(fileList, 'filename').join('\n')
-            )
-            .value();
-          if (repeatedIndexes.length > 0) {
-            return callback(
-              new Error(`Repeated migration indexes:\n${repeatedIndexes.join('\n')}`)
-            );
+        }
+
+        // In the past, we uniquely identified and ordered migrations by an integer index.
+        // However, this frequently causes conflicts between branches. To avoid this, we
+        // switched to identifying migrations by a timestamp (e.g. `20220810085513`).
+        //
+        // However, we need to actually migrate from using an integer index to
+        // using a timestamp. We'll achieve this using a multi-step deploy.
+        //
+        // First, we'll deploy a version of the code that still has the index embedded
+        // in the filename. We'll use that to sync the new timestamps to the migrations
+        // table.
+        //
+        // Only after that will we deploy a version that removes the index from the
+        // migration filenames. As part of that, we'll add some safety checks that
+        // will ensure that if we're deploying a migrations directory that doesn't
+        // have *any* indexes in the filenames, we'll error out and prompt the user
+        // to deploy an earlier version that still has timestamps.
```

Comment:
We could have also done this matching by looking at the text portion of the migration filename, but I couldn't be 100% sure that we never changed the names of migrations after they were applied, so that felt less robust. We *could* validate that there's a 1:1 mapping from existing files (the text portion) to the each entry in the `filename` column in the database and error if that's not the case, giving the operator an opportunity to manually fix the filename in the database. Thoughts?

Reply:
Glad you remembered that we've don this! The current strategy in this PR is "auto-fix based on index when we first switch to timestamps", so I'll stick with that.